flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/functions/DeclaringAsyncKeyedCoProcessFunction.java

```diff
@@ -0,0 +1,170 @@
+import org.apache.flink.runtime.asyncprocessing.declare.DeclarationContext;
+import org.apache.flink.runtime.asyncprocessing.declare.DeclarationException;
+import org.apache.flink.streaming.api.TimeDomain;
+import org.apache.flink.streaming.api.TimerService;
+import org.apache.flink.streaming.api.functions.co.CoFlatMapFunction;
+import org.apache.flink.streaming.api.functions.co.KeyedCoProcessFunction;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.function.ThrowingConsumer;
+
+/**
+ * A function that processes elements of two keyed streams and produces a single output one.
+ *
+ * <p>The function will be called for every element in the input streams and can produce zero or
+ * more output elements. Contrary to the {@link CoFlatMapFunction}, this function can also query the
+ * time (both event and processing) and set timers, through the provided {@link Context}. When
+ * reacting to the firing of set timers the function can emit yet more elements.
+ *
+ * <p>An example use-case for connected streams would be the application of a set of rules that
+ * change over time ({@code stream A}) to the elements contained in another stream (stream {@code
+ * B}). The rules contained in {@code stream A} can be stored in the state and wait for new elements
+ * to arrive on {@code stream B}. Upon reception of a new element on {@code stream B}, the function
+ * can now apply the previously stored rules to the element and directly emit a result, and/or
+ * register a timer that will trigger an action in the future.
+ *
+ * @param <K> Type of the key.
+ * @param <IN1> Type of the first input.
+ * @param <IN2> Type of the second input.
+ * @param <OUT> Output type.
+ */
+@Internal
+public abstract class DeclaringAsyncKeyedCoProcessFunction<K, IN1, IN2, OUT>
+        extends KeyedCoProcessFunction<K, IN1, IN2, OUT> {
+
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * This method is called for each element in the first of the connected streams.
+     *
+     * <p>This function can output zero or more elements using the {@link Collector} parameter and
+     * also update internal state or set timers using the {@link Context} parameter.
+     *
+     * @param value The stream element
+     * @param ctx A {@link Context} that allows querying the timestamp of the element, querying the
+     *     {@link TimeDomain} of the firing timer and getting a {@link TimerService} for registering
+     *     timers and querying the time. The context is only valid during the invocation of this
+     *     method, do not store it.
+     * @param out The collector to emit resulting elements to
+     * @throws Exception The function may throw exceptions which cause the streaming program to fail
+     *     and go into recovery.
+     */
+    public void processElement1(IN1 value, Context ctx, Collector<OUT> out) throws Exception {
+        throw new IllegalAccessException("This method is replaced by declareProcess1.");
```

Comment:
what does this exception message mean?

Reply:
This method is inherited from KeyedCoProcessFunction, but we want to use `declareProcess1` to replace it. 
This method should not be called as expected, so an exception thrown to prevent misuse.